odoo/models.py

```diff
@@ -5317,17 +5313,98 @@ def mapped(self, func):
             return self                 # support for an empty path of fields
         if isinstance(func, str):
             recs = self
-            for name in func.split('.'):
-                recs = recs._fields[name].mapped(recs)
+            for field in self._resolve_field_sequence(func):
+                recs = field.mapped(recs)
             return recs
-        else:
+        elif callable(func):
             return self._mapped_func(func)
+        else:
+            raise TypeError("Expected a function or a dot-separated sequence of field names")
+
+    def _filtered_domain_func(self, domain):
+        """Compile the domain as a function
+
+        :param domain: :ref:`A search domain <reference/orm/domains>`.
+        :return: a function ``f`` such that ``f(record)`` returns True if ``record``
+            satisfies the ``domain``.
+        """
+        if not domain:
+            return expression.TRUE_FUNCTION
+
+        def make_leaf_fn(leaf):
+            key, operator, value = expression.normalize_leaf(leaf)
+            # Optimization: short-circuit operator
+            if operator == "=?" and not value:
```

Comment:
This optimisation short-curcuits to `TRUE_FUNCTION` before the field in the left operand is even read.
It's only applicable in case the right operand is `False`.

Without this, the outcome is the same but the field is read (`mapped`) and that can have a big impact in performance (it may trigger sql queries, or computations)